src/components/SwapWidget/useSwapWidget.ts

```diff
@@ -0,0 +1,254 @@
+    sourceChain: formValues.sourceChain,
+    routeLoading,
+    numberOfTransactions: numberOfTransactions ?? 0,
+    route: routeResponse,
+    insufficientBalance,
+  };
+}
+
+export interface FormValues {
+  amountIn: string;
+  sourceChain?: Chain;
+  sourceAsset?: AssetWithMetadata;
+  destinationChain?: Chain;
+  destinationAsset?: AssetWithMetadata;
+}
+
+// useFormValues returns a set of form values that are used to populate the swap widget
+// and handles logic regarding setting initial values based on local storage and other form values.
+function useFormValues() {
+  const { chains } = useChains();
+  const { assetsByChainID, getFeeDenom } = useAssets();
+
+  const [formValues, setFormValues] = useState<FormValues>({
+    amountIn: "",
+  });
+
+  // Select initial source chain.
+  // - If chainID exists in local storage, use that.
+  // - Otherwise, default to cosmoshub-4.
+  useEffect(() => {
+    if (!formValues.sourceChain && chains.length > 0) {
+      const chainID =
+        localStorage.getItem("IBC_DOT_FUN__LAST_SOURCE_CHAIN") ?? "cosmoshub-4";
+      setFormValues((values) => ({
+        ...values,
+        sourceChain: chains.find((chain) => chain.chain_id === chainID),
+      }));
+    }
+  }, [chains, formValues.sourceChain]);
+
+  // When source chain changes, save to local storage.
+  useEffect(() => {
+    if (formValues.sourceChain) {
+      localStorage.setItem(
+        "IBC_DOT_FUN__LAST_SOURCE_CHAIN",
+        formValues.sourceChain.chain_id
+      );
+    }
+  }, [formValues.sourceChain]);
+
+  // Select initial source asset.
+  // - If fee denom exists for source chain, use that.
+  // - Otherwise, default to first asset in list.
+  useEffect(() => {
+    if (formValues.sourceChain && !formValues.sourceAsset) {
+      const feeAsset = getFeeDenom(formValues.sourceChain.chain_id);
+
+      if (feeAsset) {
+        setFormValues((values) => ({
+          ...values,
+          sourceAsset: feeAsset,
+        }));
+      } else {
+        const assets = assetsByChainID(formValues.sourceChain.chain_id);
+        if (assets.length > 0) {
+          setFormValues((values) => ({
+            ...values,
+            sourceAsset: assets[0],
+          }));
+        }
+      }
+    }
+  }, [
+    assetsByChainID,
+    formValues.sourceAsset,
+    formValues.sourceChain,
+    getFeeDenom,
+  ]);
+
+  // If destination chain is defined, but no destination asset, select an initial destination asset.
+  // - If fee denom exists for destination chain, use that.
+  // - Otherwise, default to first asset in list.
+  useEffect(() => {
+    if (formValues.destinationChain && !formValues.destinationAsset) {
+      const feeAsset = getFeeDenom(formValues.destinationChain.chain_id);
+
+      if (feeAsset) {
+        setFormValues((values) => ({
+          ...values,
+          destinationAsset: feeAsset,
+        }));
+      } else {
+        const assets = assetsByChainID(formValues.destinationChain.chain_id);
```

Comment:
Same comment I made for source assets. Also feels like this logic could be abstracted to a helper function since it's duplicated